torch/autograd/gradcheck.py

```diff
@@ -806,41 +886,88 @@ def wrapped_fn(*inputs):
         gradcheck_fn(real_fn, real_func_out, tupled_inputs, real_outputs, eps,
-                     rtol, atol, check_grad_dtypes, nondet_tol, complex_indices)
+                     rtol, atol, check_grad_dtypes, nondet_tol, complex_indices=complex_indices)
     else:
         gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
                      rtol, atol, check_grad_dtypes, nondet_tol)
 
+    complex_indices = [i for i, inp in enumerate(tupled_inputs) if is_tensor_like(inp) and inp.is_complex()]
+    if check_forward_ad:
+        if complex_indices:
+            real_fn, imag_fn = _real_and_imag_input(func)
+
+            imag_inputs = [inp.imag if inp.is_complex() else inp for inp in tupled_inputs]
+            imag_func_out = imag_fn(*imag_inputs)
+            diff_imag_func_out = _differentiable_outputs(imag_func_out)
+            gradcheck_fn(imag_fn, imag_func_out, imag_inputs, diff_imag_func_out, eps,
+                         rtol, atol, check_grad_dtypes, nondet_tol,
+                         complex_indices=complex_indices, test_imag=True, use_forward_ad=True)
+
+            real_inputs = [inp.real if inp.is_complex() else inp for inp in tupled_inputs]
+            real_func_out = real_fn(*real_inputs)
+            diff_real_func_out = _differentiable_outputs(real_func_out)
+            gradcheck_fn(real_fn, real_func_out, real_inputs, diff_real_func_out, eps,
+                         rtol, atol, check_grad_dtypes, nondet_tol, complex_indices=complex_indices,
+                         use_forward_ad=True)
+        else:
+            gradcheck_fn(func, func_out, tupled_inputs, outputs, eps,
+                         rtol, atol, check_grad_dtypes, nondet_tol, use_forward_ad=True)
+
 
-def _slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
-                    atol, check_grad_dtypes, nondet_tol, complex_indices=None, test_imag=False):
+def _slow_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol, atol, check_grad_dtypes,
+                    nondet_tol, *, use_forward_ad=False, complex_indices=None, test_imag=False):
     if not outputs:
         return _check_no_differentiable_outputs(func, tupled_inputs, _as_tuple(func_out), eps)
 
     numerical = _transpose(_get_numerical_jacobian(func, tupled_inputs, outputs, eps=eps))
 
-    for i, o in enumerate(outputs):
-        analytical = _check_analytical_jacobian_attributes(tupled_inputs, o, nondet_tol, check_grad_dtypes)
-        inp_tensors = _iter_tensors(tupled_inputs, True)
+    if use_forward_ad:
+        complex_indices = [i for i, inp in enumerate(tupled_inputs) if inp.is_complex()]
+        analytical_forward = _get_analytical_jacobian_forward_ad(func, tupled_inputs, outputs)
+
+        for i, (n_per_out, a_per_out) in enumerate(zip(numerical, analytical_forward)):
+            for j, (n, a) in enumerate(zip(n_per_out, a_per_out)):
+                if a.numel() != 0 or n.numel() != 0:
```

Comment:
Shouldn't torch.allclose be able to compare tensors that have 0 numel? Is this check redundant? (a comment would be nice if there is a reason)

Reply:
I just re-used existing code here. Not sure.

Reply:
Oh I think it can be removed. I originally added it only for debugging purposes, when I wanted to compute `(a - n).abs().max()`, and reduction ops require non-zero numel.

Reply:
Updated.